src/sagas/options.js

```diff
@@ -0,0 +1,228 @@
+import { all, put, select, takeLatest, takeEvery, call } from 'redux-saga/effects'
+
+import { saveSSHKey as saveSSHKeyAction, saveUserOptionsOnBackend, setOption, setSSHKey, loadingUserOptionsInProgress, loadingUserOptionsFinished, stopSchedulerForResumingNotifications, startSchedulerForResumingNotifications } from '_/actions'
+import { callExternalAction } from './utils'
+import { processUser } from './index'
+
+import {
+  GET_SSH_KEY,
+  SAVE_GLOBAL_OPTIONS,
+  SAVE_SSH_KEY,
+  SAVE_VMS_OPTIONS,
+} from '_/constants'
+
+import type { SaveGlobalOptionsActionType, SaveVmsOptionsActionType, VmSettingsType } from '_/ovirtapi/types'
+import { fromJS } from 'immutable'
+
+function* fetchSSHKey (action: Object): any {
+  const result = yield * callExternalAction('getSSHKey', Api.getSSHKey, action)
+  if (result.error) {
+    return
+  }
+  if (result.ssh_public_key && result.ssh_public_key.length > 0) {
+    yield put(setSSHKey(Api.SSHKeyToInternal({ sshKey: result.ssh_public_key[0] })))
+  } else {
+    yield put(setSSHKey(Api.SSHKeyToInternal({ sshKey: '' })))
+  }
+}
+
+export function* saveSSHKey ([ name, value ]: any): any {
+  if (value === undefined) {
+    // should be possible to clear ssh key by setting to empty string
+    return { changes: [] }
+  }
+  const sshId = yield select((state) => state.options.getIn(['ssh', 'id']))
+  const currentKey = yield select((state) => state.options.getIn(['ssh', 'key']))
+  if (currentKey === value) {
+    return { changes: [name], sameAsCurrent: true }
+  }
+  const userId = yield select((state) => state.config.getIn(['user', 'id']))
+  const result = yield * callExternalAction(
+    'saveSSHKey',
+    Api.saveSSHKey,
+    saveSSHKeyAction({ sshId, key: value, userId }),
+    true)
+
+  // if no entry in user_profiles exist then the response is simply {status: complete}
+  // without the content (ssh key body). In order to avoid extra call we set
+  // the value that we have just sent to the server. If the call was successfull the value is correct,
+  // if the call failed we expect an error flag to be present anyway.
+  return { content: value, ...result, changes: [name] }
+}
+
+function mergeWithDefaults (currentOptions: Object, paths: Array<Array<string>>, defaultOptions?: Object): any {
+  if (defaultOptions === undefined) {
+    return undefined
+  }
+  const deafultOptionsTemplate = fromJS(defaultOptions)
+  const defaultOptionsForAllPaths = paths.reduce((acc, path) => acc.setIn(path, deafultOptionsTemplate), fromJS({}))
+
+  return currentOptions.mergeDeepWith((prev, next, key) => {
+    if (next === undefined) {
+      console.error(`Unknown property ${key} has no default value`)
+    }
+    if (prev !== undefined) {
+      // use existing property if present
+      return prev
+    }
+    // use default
+    return next
+  }, defaultOptionsForAllPaths)
+}
+
+function* saveRemoteOptions (newOptions: Object, paths: Array<Array<string>>, defaultOptions?: Object): any {
```

Comment:
I'm having a difficult time figuring the purpose of this saga.

I know there is a hierarchy order for how things are being applied, but it is getting lost in the implementation.

I suggest putting in a bit of comments about how the server side options are being combine with the user changes and default options.  I would find it useful so I can keep in mind what is intended, the actual and the data flow.